### ISSUE TITLE
Inconsistent position of 'const' between immutable targets of pointers and slices

### ISSUE THREAD

**Issue by dolmen** [romshark/Go-1-2-Proposal---Immutability#16]
In Go, slices are just iterable pointers with safe bounds (safe thanks to runtime enforcement).
The type declaraction is consistent: the derefencing is on the left of the element type.

Here is [some Go code](https://play.golang.org/p/Pe26moO2WH_t) that use mutability to indirectly modify the same memory location first through a pointer, then through a slice:
```
package main

import (
	"fmt"
)

func main() {
	ints := [1]int{42}

	ptr := &ints[0]
	fmt.Println(*ptr)
	*ptr++
	fmt.Println(*ptr)

	slc := ints[:1]
	fmt.Println(slc[0])
	slc[0]++
	fmt.Println(slc[0])

	fmt.Println(ints[0])
}
```

According to this proposal, to forbid modification through `*ptr` or `slc[0]` (`*ptr++` and `slc[0]++` should be rejected at compile time), we have to declare like this:
```
ptr := * const int(&ints[0])
slc := const []int(ints[:1])
``` 
References:
- [2.8.2 Dereferencing a pointer](https://github.com/romshark/Go-1-2-Proposal---Immutability/blob/ffebd19531bf6a091360dfb63dc05741d4b9f944/README.md#282-dereferencing-a-pointer)
- [2.7 Slice aliasing](https://github.com/romshark/Go-1-2-Proposal---Immutability/blob/ffebd19531bf6a091360dfb63dc05741d4b9f944/README.md#27-slice-aliasing)

I argue that the position of the `const` keyword in the type types expression are inconsistent: the `const` is on the right of `*` but on the left of `[]`.

This is a follow-up of my tweets at https://twitter.com/omengue/status/1047743716461629440

**Comment by dolmen** [comment]
It looks like the inconsistency has been noticed as the entries of the FAQ seem to use instead a syntax for immutability of slice elements that is consistent with immutability of pointer dereferencing.
So at least section 1 and 2 should be fixed.

**Comment by romshark** [comment]
According to [2.8 Address Operators](https://github.com/romshark/Go-1-2-Proposal---Immutability#28-address-operators) and [2.7 Slice Aliasing](https://github.com/romshark/Go-1-2-Proposal---Immutability#27-slice-aliasing) the results will be the following:
```go
var ints const [] const int = []{1, 2, 3}

ptr := &ints[0] // * const int
i := *ptr // const int
*ptr++ // Compile-time error

slc := ints[:1] // const [] const int
i = slc[0] // const int
slc[0]++ // Compile-time error
```

**Comment by romshark** [comment]
Please consider closing this issue if you consider it resolved

**Comment by dolmen** [comment]
My point is not how to change the type of `ints`. That's cheating.



In [1.1.4](https://github.com/romshark/Go-1-2-Proposal---Immutability#114-inconsistent-concept-of-constants), two `const` are needed for full immutability:
```go
var each2 = const [] const byte{'e', 'a', 'c', 'h'}
```

In [1.2.1](https://github.com/romshark/Go-1-2-Proposal---Immutability#121-safe-and-precise-code) this question is quite misleading:
> Will it mutate any T referenced by any item of v?
>
>    No, it can't. The Ts referenced by any item of v are immutable
Here is an expression that will modify the array pointed by the `v` slice:
```go
v[0] = new(const T)
```
The type of `v` should be changed into `[] const * const T` to really ensure safety.

In [2.3](https://github.com/romshark/Go-1-2-Proposal---Immutability#23-immutable-arguments), `ReadObj` is forbidden from modifying `*obj` and `obj.MutableField`, but nothing stops it from modifying `obj.MutableField.MutableField` because `obj.MotableField` is of type `*Object`. This is quite misleading about the immutability power of this proposal. It has the same flaws as `const` in C/C++.

**Comment by merhalak** [comment]
In C/C++ const modifier must be after pointer, but may be before type.

Such as (from wiki: https://en.wikipedia.org/wiki/Const_(computer_programming)):
```
int *ptr; // *ptr is an int value
int const *ptrToConst; // *ptrToConst is a constant (int: integer value)
int * const constPtr; // constPtr is a constant (int *: integer pointer)
int const * const constPtrToConst; // constPtrToConst is a constant (pointer)
                                   // as is *constPtrToConst (value)
```

It's looks better, than const modifier before.

In Go, we write type after variable name, why we can't write modifier after type name?

**Comment by merhalak** [comment]
Writing modifier **always** after type name looks good, isn't it?

**Comment by romshark** [comment]
> My point is not how to change the type of ints. That's cheating.
> 
> In 1.1.4, two const are needed for full immutability:
> 
> ```go
> var each2 = const [] const byte{'e', 'a', 'c', 'h'}
> ```
> 
> In 1.2.1 this question is quite misleading:
> 
>> Will it mutate any T referenced by any item of v?
>>
>> No, it can't. The Ts referenced by any item of v are immutable
>> Here is an expression that will modify the array pointed by the v slice:
> 
> ```go
> v[0] = new(const T)
> ```
>
> The type of v should be changed into [] const * const T to really ensure safety.

That's all correct. In fact, most of the examples including slices and maps in the document are wrong because I initially falsely assumed that `&[]` is just a combination of the _index_ and _reference_ operators, because if it would be just that then `&` would take the address of the copy returned from `[]` located on the stack and not the address of the actual slice/map item. But Go seems to do some language-level magic making `&[]` one separate operator taking the address of the actual items (https://goplay.space/#PYb1K0nmmTL).

With a `const` qualifier it will really become a nightmare of `const [] const * const T` and that's why in the [second revision](https://github.com/romshark/Go-1-2-Proposal---Immutability/milestone/2) I'll introduce [mutability qualification propagation](https://github.com/romshark/Go-1-2-Proposal---Immutability/issues/20). Eventually the example will be:
```go
var each2 = []byte{'e', 'a', 'c', 'h'}
```
If immutability will be default behavior, or if not then:
```go
var each2 = immut []byte{'e', 'a', 'c', 'h'}
```

> In [2.3](https://github.com/romshark/Go-1-2-Proposal---Immutability#23-immutable-arguments), `ReadObj` is forbidden from modifying `*obj` and `obj.MutableField`, but nothing stops it from modifying `obj.MutableField.MutableField` because `obj.MotableField` is of type `*Object`. This is quite misleading about the immutability power of this proposal. It has the same flaws as `const` in C/C++.

Wrong. Even though `obj.MutableField` is of a mutable type, it's part of `obj` which is an immutable type `const Object` and is thus **contextually immutable**. Subsequently, `obj.MutableField.MutableField` is immutable too, because it's part of a (contextually) immutable type, which is part of an immutable type.

Having immutable composite types that don't protect their fields would make the entire concept useless similar to JavaScript `const` references because it wouldn't protect from mutable shared state.

I think I should describe **contextual immutability** better (or did I miss that point entirely?!).

**Comment by romshark** [comment]
@merhalak 
In Go - types are read and written from left to right in a (western) human-friendly way.

```go
// some absurdly complex type:
**[]map[int]*T
```
This is a **pointer** to a **pointer** to a **slice** of **maps** of **integers** to **pointers** to **T**. It's much easier to read than any C/C++ equivalent.

Mutability qualifiers should be no exception to this rule:

```go
// an even more absurd one:
mut * immut * mut [] immut map[mut int] mut * immut T
```
This is a **mutable** pointer to an **immutable** pointer to a **mutable** slice of **immutable** maps of **mutable** integers to **mutable** pointers to **immutable** T.

A mutability qualifier always affects the type(s) on the right and propagates until it's canceled out ([MQP](#20)).

Remember, those were just absurdly complicated examples, nobody will ever write things like this (people usually get hanged for doing so). You'll most likely se no more than:
```go
mut [] immut *T
```
Which is a **mutable slice** of **immutable pointers** to **immutable T** (because immut propagates due to [MQP](#20)).

**Comment by merhalak** [comment]
@romshark thanks for your explanation. Sorry for off topic.
